src/engraving/dom/noteline.cpp

```diff
@@ -0,0 +1,250 @@
+{
+    NoteLineSegment* seg = new NoteLineSegment(this, parent);
+    seg->setTrack(track());
+    return seg;
+}
+
+PropertyValue NoteLine::propertyDefault(Pid propertyId) const
+{
+    switch (propertyId) {
+    case Pid::BEGIN_TEXT:
+    case Pid::CONTINUE_TEXT:
+    case Pid::END_TEXT:
+        return "";
+    case Pid::LINE_VISIBLE:
+        return true;
+    case Pid::BEGIN_TEXT_OFFSET:
+    case Pid::CONTINUE_TEXT_OFFSET:
+    case Pid::END_TEXT_OFFSET:
+        return PointF(0, 0);
+    case Pid::BEGIN_HOOK_TYPE:
+    case Pid::END_HOOK_TYPE:
+        return HookType::NONE;
+    case Pid::BEGIN_TEXT_PLACE:
+    case Pid::CONTINUE_TEXT_PLACE:
+    case Pid::END_TEXT_PLACE:
+        return TextPlace::LEFT;
+    case Pid::BEGIN_HOOK_HEIGHT:
+    case Pid::END_HOOK_HEIGHT:
+        return Spatium(1.5);
+    case Pid::PLACEMENT:
+        return PlacementV::ABOVE;
+    case Pid::DIAGONAL:
+        return true;
+    case Pid::NOTELINE_PLACEMENT:
+        return NoteLineEndPlacement::OFFSET_ENDS;
+    case Pid::OFFSET:
+        return PointF();
+    default:
+        return TextLineBase::propertyDefault(propertyId);
+    }
+}
+
+PropertyValue NoteLine::getProperty(Pid propertyId) const
+{
+    switch (propertyId) {
+    case Pid::PLACEMENT:
+        return PlacementV::ABOVE;
+    case Pid::NOTELINE_PLACEMENT:
+        return m_lineEndPlacement;
+    default:
+        return TextLineBase::getProperty(propertyId);
+    }
+}
+
+bool NoteLine::setProperty(Pid propertyId, const PropertyValue& val)
+{
+    switch (propertyId) {
+    case Pid::NOTELINE_PLACEMENT:
+        setLineEndPlacement(val.value<NoteLineEndPlacement>());
+        break;
+    default:
+        return TextLineBase::setProperty(propertyId, val);
+    }
+
+    return true;
+}
+
+void NoteLine::addLineAttachPoints()
+{
+    assert(anchor() == Spanner::Anchor::NOTE);
+
+    NoteLineSegment* frontSeg = toNoteLineSegment(frontSegment());
+    NoteLineSegment* backSeg = toNoteLineSegment(backSegment());
+    Note* startNote = nullptr;
+    Note* endNote = nullptr;
+    if (startElement() && startElement()->isNote()) {
+        startNote = toNote(startElement());
+    }
+    if (endElement() && endElement()->isNote()) {
+        endNote = toNote(endElement());
+    }
+    if (!frontSeg || !backSeg || !startNote || !endNote) {
+        return;
+    }
+    double startX = frontSeg->ldata()->pos().x();
+    double endX = backSeg->pos2().x() + backSeg->ldata()->pos().x(); // because pos2 is relative to ipos
+    // Here we don't pass y() because its value is unreliable during the first stages of layout.
+    // The y() is irrelevant anyway for horizontal spacing.
+    startNote->addLineAttachPoint(PointF(startX, 0.0), this);
+    endNote->addLineAttachPoint(PointF(endX, 0.0), this);
+}
+
+void NoteLine::reset()
+{
+    undoResetProperty(Pid::NOTELINE_PLACEMENT);
+    TextLineBase::reset();
+}
+
+NoteLine* NoteLine::createFromTextLine(TextLine* textLine)
```

Comment:
If this function is only used as part of the compatibility mapping, as I seem to understand, I think it should be a private method in CompatUtils, not a member of this class